File: backend/src/storage/audio-object-keys.ts
export const DEFAULT_TRACK_AUDIO_EXTENSION = 'm4a';

export type TrackAudioObjectKeyInput = {
  projectId: string;
  sessionId: string;
  trackId: string;
  extension?: string;
};

export function buildTrackAudioObjectKey(input: TrackAudioObjectKeyInput) {
  const extension = normalizeAudioExtension(input.extension ?? DEFAULT_TRACK_AUDIO_EXTENSION);

  return [
    'projects',
    encodeS3KeySegment(input.projectId),
    'sessions',
    encodeS3KeySegment(input.sessionId),
    'tracks',
    `${encodeS3KeySegment(input.trackId)}.${extension}`,
  ].join('/');
}

export function buildTrackAudioObjectPrefix(projectId: string, sessionId: string) {
  return [
    'projects',
    encodeS3KeySegment(projectId),
    'sessions',
    encodeS3KeySegment(sessionId),
    'tracks',
  ].join('/');
}

function encodeS3KeySegment(segment: string) {
  const trimmedSegment = segment.trim();

  if (!trimmedSegment) {
    throw new Error('S3 object key segments must not be empty.');
  }

  return encodeURIComponent(trimmedSegment);
}

function normalizeAudioExtension(extension: string) {
  const normalizedExtension = extension.trim().replace(/^\./, '').toLowerCase();

  if (!/^[a-z0-9]+$/.test(normalizedExtension)) {
    throw new Error('Audio file extension must be alphanumeric.');
  }

  return normalizedExtension;
}

File: backend-dotnet/src/Loopr.Api/Storage/AudioObjectKeys.cs
using System.Text.RegularExpressions;

namespace Loopr.Api.Storage;

public sealed record TrackAudioObjectKeyInput(
    string ProjectId,
    string SessionId,
    string TrackId,
    string? Extension = null
);

public static class AudioObjectKeys
{
    public const string DefaultTrackAudioExtension = "m4a";

    public static string BuildTrackAudioObjectKey(TrackAudioObjectKeyInput input)
    {
        var extension = NormalizeAudioExtension(
            input.Extension ?? DefaultTrackAudioExtension
        );

        return string.Join(
            "/",
            [
                "projects",
                EncodeS3KeySegment(input.ProjectId),
                "sessions",
                EncodeS3KeySegment(input.SessionId),
                "tracks",
                $"{EncodeS3KeySegment(input.TrackId)}.{extension}",
            ]
        );
    }

    public static string BuildTrackAudioObjectPrefix(string projectId, string sessionId)
    {
        return string.Join(
            "/",
            [
                "projects",
                EncodeS3KeySegment(projectId),
                "sessions",
                EncodeS3KeySegment(sessionId),
                "tracks",
            ]
        );
    }

    private static string EncodeS3KeySegment(string segment)
    {
        var trimmedSegment = segment.Trim();

        if (string.IsNullOrEmpty(trimmedSegment))
        {
            throw new InvalidOperationException("S3 object key segments must not be empty.");
        }

        return Uri.EscapeDataString(trimmedSegment);
    }

    private static string NormalizeAudioExtension(string extension)
    {
        var normalizedExtension = extension.Trim().TrimStart('.').ToLowerInvariant();

        if (!Regex.IsMatch(normalizedExtension, "^[a-z0-9]+$"))
        {
            throw new InvalidOperationException(
                "Audio file extension must be alphanumeric."
            );
        }

        return normalizedExtension;
    }
}
